[Docs] Update alert and console interactions

TAYTS opened the pull request:
### Summary

Fixes #3940: Improve documentation experiences

### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- [x] Checked in **mobile**
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- [x] Props have proper **autodocs**
~~- [] Added **[documentation](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md)**~~
- [x] Checked **[Code Sandbox](https://codesandbox.io/)** works for the any docs examples
~~- [ ] Added or updated **[jest tests](https://github.com/elastic/eui/blob/master/wiki/testing.md)**~~
~~- [ ] Checked for **breaking changes** and labeled appropriately~~
~~- [ ] Checked for **accessibility** including keyboard-only and screenreader modes~~
~~- [ ] A **[changelog](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately~~

kibanamachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

TAYTS commented:
@chandlerprall 
Hi I have done some of the implementations, sorry I was thinking to complete all of the implementations then submit the PR for review but here are some of changes implemented so far with its screenshots and code snippets.

- Added state watcher for `onToggle` in `EuiAccordion`

![Screenshot 2020-09-11 at 7 41 43 AM](https://user-images.githubusercontent.com/29126748/92825549-3edcef80-f402-11ea-897c-6f94c11d577f.png)
![Screenshot 2020-09-11 at 7 43 05 AM](https://user-images.githubusercontent.com/29126748/92825709-70ee5180-f402-11ea-97a1-3d2d71fdd834.png)

- Added page content for the `EuiHeader`

![Screenshot 2020-09-11 at 7 39 20 AM](https://user-images.githubusercontent.com/29126748/92825235-ea397480-f401-11ea-9d14-0d799b19f2e9.png)
![Screenshot 2020-09-11 at 7 40 40 AM](https://user-images.githubusercontent.com/29126748/92825425-1b19a980-f402-11ea-8162-2247ca0cf8f8.png)

- Added global toast to `EuiPanel`, `EuiButton`, `EuiContextMenu`, `EuiKeyPadMenu`, `EuiLink`, `EuiTreeView`, `EuiSideNav`, `EuiSteps`

<img width="1186" alt="Screenshot 2020-09-11 at 7 45 06 AM" src="https://user-images.githubusercontent.com/29126748/92826040-d3475200-f402-11ea-8657-15e218dc1b10.png">

- Added state watcher for the `EuiBreadcrumbs`

![Screenshot 2020-09-11 at 7 47 59 AM](https://user-images.githubusercontent.com/29126748/92826297-215c5580-f403-11ea-8ba9-751bd1251a11.png)
![Screenshot 2020-09-11 at 7 48 18 AM](https://user-images.githubusercontent.com/29126748/92826331-2ae5bd80-f403-11ea-8ef9-7f022203b8a7.png)

- Stop link redirect for `EuiBreadcrumbs`, `EuiButton`, `EuiContextMenu`, `EuiKeyPadMenu`, `EuiLink`, `EuiSideNav`, `EuiTabs`
``` Javascript
const stopRedirect = e => {
    e.preventDefault();
}

<EuiButton href="http://www.elastic.co" onClick={stopRedirect}>Link to elastic.co</EuiButton>
```

- Added a global toast component which expose the `addToast` method

<img width="584" alt="Screenshot 2020-09-11 at 8 12 34 AM" src="https://user-images.githubusercontent.com/29126748/92829228-8d8c8880-f406-11ea-901a-883f4d8ccf9e.png">

- Replace `window.alert()` for table row size in `EuiPagination` with state update

![Screenshot 2020-09-11 at 8 14 26 AM](https://user-images.githubusercontent.com/29126748/92829456-d17f8d80-f406-11ea-9e7b-a4134bdb0c56.png)

cchaos commented:
Jenkins, test this

cchaos commented:
Hey @TAYTS ! Nice large PR you've started 😂 

A couple of feedback items from a documentation standpoint:

The global toast list is a great way to surface callback functions that return something other than the simple handling event. For instance, if the event passes back something other than `MouseEvent<HTMLButtonElement>` which tends to happen mostly in more complicated components. But for simple `onClick()` functions where we just need to add it to show the hover and focus states, like on EuiButton, let's just return an empty function like `onClick: () => {}`.

There are a couple reasons for this:

1. It keeps the `DemoJS` tab code simplified to where a consumer can copy/paste the whole example and it comes with everything it needs to work out of the box without having to remove Docs-only code.
2. The CodeSandbox links will break with the global toast service which is importing something that is Docs-only and isn't shipped with the distributed version of EUI. 

I'll then pass to @chandlerprall to evaluate the code of the new toast service.

TAYTS commented:
@cchaos For the components with `#` or empty value for the `href` prop, should we use `e.preventDefault()` together with the `onClick` or simply set it to the current page path to prevent the page redirect?

Since there are some demos which are solely just to show the `href` prop (e.g. `<EuiBreadcrumb />`, `<EuiLink />` ), will adding the `e.preventDefault()` complicate the examples or it would be good to show to combination of `href` and `onClick`?

cchaos commented:
As far as I can tell an empty `href=""` doesn't navigate anywhere while still showing up as an `<a>` tag (at least locally). I think that would be the best solution without needing to remember to add `preventDefault` to every usage.

TAYTS commented:
@cchaos I think some components does not render `<a>` if empty value is provided (tested on `<EuiButton>` and `<EuiBreadcrumbs>`), with this is adding current page path ok? (e.g. `href: '#/navigation/breadcrumbs'` for the `<Euibreadcrumbs>`)

thompsongl commented:
One solution is to handle `href="#"` more globally, for instance:

Adding the following to `componentDidMount` in `src-docs/src/views/app_view.js` prevents location changes across the site, without the need for `onClick` in every component:

```js
document.addEventListener('click', e => {
  if (e.target.getAttribute('href') === '#') {
    e.preventDefault();
  }
});
```

TAYTS commented:
@thompsongl 
I have tried the suggested method but it would not work all the time as the event target might be the children of the component which does not have the `href` attribute.

cchaos commented:
Hey @TAYTS thanks for taking on this chore for us. There seems to be quite a few more additions outside of the original ask from the issue being applied here. For instance the addition of `target="_blank"`. We try not to open new tabs because it is bad for accessibility and instead we just rely on ensuring the users can use the browser back button to return.

Also the addition of the Page content in the header examples. I'm not sure why this addition is necessary.

It would probably be most helpful to scope this PR to the original task of removing `window.alert`s and `console.log`s which will also make it simpler and quicker to review. Then follow up with suggestion on other improvements in another PR or issue to discuss.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4028/

TAYTS commented:
@cchaos I have made the changes as suggested, please help to review and let me know if there is any parts required to modify. Thank you.

cchaos commented:
Jenkins, test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4028/

cchaos commented on src-docs/src/views/suggest/global_filter_form.js:
```diff
@@ -147,6 +147,7 @@ const GlobalFilterForm = (props) => {
         <EuiFlexItem style={{ maxWidth: '188px' }}>
           <EuiFormRow label="Field">
             <EuiComboBox
+              isDisabled={useCustomLabel}
```
This logic isn't quite right. The custom label is not dependent nor does it dictate how the rest of the form behaves. The switch `useCustomLabel` of should only show/hide the Custom label input. 
```suggestion
```

TAYTS replied:
OK

cchaos reviewed (changes requested):
There are a few spots where the alerts or console logs are still needed to help the user understand the callback. Also, it looks like there are still some issues from master with the updated linter.

TAYTS commented on src-docs/src/views/datagrid/datagrid.js:
```diff
@@ -219,9 +219,7 @@ export default () => {
-      onColumnResize={eventData => {
-        console.log(eventData);
-      }}
+      onColumnResize={eventData => {}}
```
OK

TAYTS commented on src-docs/src/views/datagrid/footer_row.js:
```diff
@@ -128,9 +128,7 @@ export default () => {
-          onColumnResize={eventData => {
-            console.log(eventData);
-          }}
+          onColumnResize={eventData => {}}
```
OK

TAYTS commented on src-docs/src/views/drag_and_drop/drag_and_drop_bare.js:
```diff
@@ -18,11 +18,9 @@ const makeList = (number, start = 1) =>
+
   return (
-    <EuiDragDropContext onDragEnd={onDragEnd}>
+    <EuiDragDropContext onDragEnd={() => {}}>
```
OK

TAYTS commented on src-docs/src/views/elastic_charts/pie_slices.js:
```diff
@@ -100,8 +100,6 @@ export default () => {
       setSliceOrderConfig({});
       setSliceOrderConfigText('');
-    } else {
-      console.warn("Couldn't find the right slice order type");
```
OK

TAYTS commented on src-docs/src/views/elastic_charts/theming_categorical.js:
```diff
@@ -116,7 +116,6 @@ export const Categorical = () => {
         setChartType('LineSeries');
         break;
       default:
-        console.warn("Couldn't find the right chart type");
```
OK

TAYTS commented on src-docs/src/views/super_date_picker/super_date_picker.js:
```diff
@@ -38,8 +38,6 @@ export default () => {
     return new Promise(resolve => {
       setTimeout(resolve, 100);
-    }).then(() => {
-      console.log(start, end, refreshInterval);
```
OK

TAYTS commented on src-docs/src/views/super_date_picker/super_date_picker_config.js:
```diff
@@ -38,8 +38,6 @@ export default () => {
     return new Promise(resolve => {
       setTimeout(resolve, 100);
-    }).then(() => {
-      console.log(start, end, refreshInterval);
```
OK

TAYTS commented on src-docs/src/views/super_date_picker/super_date_picker_custom_quick_select.js:
```diff
@@ -49,8 +49,6 @@ export default () => {
     return new Promise(resolve => {
       setTimeout(resolve, 100);
-    }).then(() => {
-      console.log(start, end, refreshInterval);
```
OK

TAYTS commented on src-docs/src/views/accessibility/keyboard_accessible.js:
```diff
@@ -12,35 +12,27 @@ export default () => (
     <EuiText>
       <EuiKeyboardAccessible>
-        <div onClick={() => window.alert('Div clicked')}>Click this div</div>
+        <div onClick={() => {}}>Click this div</div>
```
OK

TAYTS commented on src-docs/src/views/datagrid/datagrid.js:
```diff
@@ -224,9 +224,7 @@ export default () => {
-      onColumnResize={(eventData) => {
-        console.log(eventData);
-      }}
+      onColumnResize={(eventData) => {}}
```
OK

TAYTS commented on src-docs/src/views/datagrid/footer_row.js:
```diff
@@ -133,9 +133,7 @@ export default () => {
-          onColumnResize={(eventData) => {
-            console.log(eventData);
-          }}
+          onColumnResize={(eventData) => {}}
```
OK

TAYTS commented on src-docs/src/views/accordion/accordion_callback.js:
```diff
@@ -1,21 +1,30 @@
+          <strong>EuiAccordion</strong> is now{' '}
+          <EuiCode>{isOpen ? 'open' : 'closed'}</EuiCode>.
         </p>
       </EuiText>
```
OK

TAYTS commented on src-docs/src/views/outside_click_detector/outside_click_detector.js:
```diff
@@ -32,6 +50,14 @@ export default () => {
+
+      <EuiSpacer size="l" />
+
+      {showCallOut ? (
```
OK

TAYTS commented on src-docs/src/views/code_editor/code_editor.js:
```diff
@@ -26,9 +26,7 @@ export default () => {
-      onBlur={() => {
-        console.log('blur');
-      }} // eslint-disable-line no-console
+      onBlur={() => {}} // eslint-disable-line no-console
```
OK

TAYTS commented on src-docs/src/views/suggest/global_filter_form.js:
```diff
@@ -99,7 +99,8 @@ const GlobalFilterForm = (props) => {
 
   const onCustomLabelSwitchChange = (e) => {
-    setCustomLabel(e.target.checked);
+    resetForm();
```
OK

cchaos commented:
Jenkins, test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4028/

cchaos reviewed (approved):
Thanks @TAYTS ! LGTM!
